server/bleep/src/webserver/answer.rs

```diff
@@ -94,34 +98,141 @@ pub async fn handle(
     }
 
-    let res = reqwest::Client::new()
-        .post(format!("{}/q", app.config.answer_api_base))
-        .json(&api::Request {
-            query: params.q,
-            user_id: params.user_id,
-            snippets: snippets.clone(),
-        })
-        .send()
+    let answer_api_client = AnswerAPIClient::new(
+        &format!("{}/q", app.config.answer_api_base),
+        &target,
+        &params.user_id,
+    );
+
+    let res1 = answer_api_client
+        .stage_1(&snippets)
+        .await
+        .map_err(|e| super::error(ErrorKind::Internal, e.to_string()))?
+        .index;
+
+    let processed_snippet = &snippets[res1];
+
+    let res2 = answer_api_client
+        .stage_2(processed_snippet)
         .await
-        .map_err(|e| {
-            super::error(
-                ErrorKind::Internal,
-                format!("failed to make request to answer API: {}", e),
-            )
-        })?;
-
-    let selection: api::Response = res.json().await.map_err(|e| {
-        super::error(
-            ErrorKind::Internal,
-            format!("answer API was not able to create a valid result: {}", e),
-        )
-    })?;
-
-    let selected_snippet = snippets.remove(selection.data.index as usize);
+        .map_err(|e| super::error(ErrorKind::Internal, e.to_string()))?
+        .answer;
+
+    // reorder snippets
+    let selected_snippet = snippets.remove(res1);
     snippets.insert(0, selected_snippet);
 
     Ok::<_, Json<super::Response<'static>>>(Json(super::Response::Answer(AnswerResponse {
         snippets,
-        selection,
+        selection: api::Response {
+            data: api::DecodedResponse {
+                index: res1 as u32,
+                answer: res2,
+            },
+            id: params.user_id,
+        },
     })))
 }
+
+#[derive(serde::Serialize)]
+struct OpenAIRequest {
+    prompt: String,
+    user_id: String,
+}
+
+struct AnswerAPIClient {
+    client: reqwest::Client,
+    host: String,
+    query: String,
+    user_id: String,
+}
+
+impl AnswerAPIClient {
+    fn new(host: &str, query: &str, user_id: &str) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            host: host.to_owned(),
+            query: query.to_owned(),
+            user_id: user_id.to_owned(),
+        }
+    }
+
+    async fn send(&self, prompt: String) -> Result<reqwest::Response, reqwest::Error> {
+        self.client
+            .post(self.host.as_str())
+            .json(&OpenAIRequest {
+                prompt: prompt.clone(),
+                user_id: self.user_id.clone(),
+            })
+            .send()
+            .await
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct Stage1Response {
+    index: usize,
+}
+
+#[derive(serde::Deserialize)]
+struct Stage2Response {
+    answer: String,
+}
+
+const DELIMITER: &str = "######";
+impl AnswerAPIClient {
+    async fn stage_1(&self, snippets: &[api::Snippet]) -> Result<Stage1Response, reqwest::Error> {
```

Comment:
Let's call this `select_snippet` (and rename the corresponding `Response` struct)